helix-term/src/application.rs

```diff
@@ -65,6 +66,7 @@ pub struct Application {
     compositor: Compositor,
     terminal: Terminal,
     pub editor: Editor,
+    registers: Registers,
```

Comment:
> Registers moves from a field on Editor to fields on the Context types to avoid borrowing conflicts.

This is because the trait takes `&mut Editor` on write and push right? And that's only needed for `editor.clipboard_provider`? Maybe we could instead move `clipboard_provider` to Application

Reply:
Even if we downgrade to a shared reference there are some borrowing conflicts on write/push:

```
error[E0502]: cannot borrow `cx.editor.registers` as mutable because it is also borrowed as immutable
    --> helix-term/src/commands.rs:2333:27
     |
2333 |         if let Err(err) = cx.editor.registers.write(reg_name, cx.editor, values) {
     |                           ^^^^^^^^^^^^^^^^^^^^-----^^^^^^^^^^^---------^^^^^^^^^
     |                           |                   |               |
     |                           |                   |               immutable borrow occurs here
     |                           |                   immutable borrow later used by call
     |                           mutable borrow occurs here
```

since `write`/`push` take `&mut self` and `&Editor`. We could probably resolve this if we restricted special registers to take only `&self` and `&Editor` but then they become less powerful. For example we couldn't store the last value written to the clipboards which we use here to retain the selections.

Kakoune's `Registers` equivalent "RegisterManager" is unfortunately not very rusty: it's a singleton, see [here](https://github.com/mawww/kakoune/blob/e3122ab2c19248b590cb0143610439ec050766b7/src/register_manager.hh#L118) and [here](https://github.com/mawww/kakoune/blob/e3122ab2c19248b590cb0143610439ec050766b7/src/utils.hh#L11-L48).

Reply:
It would be nice to move the clipboard provider out of Editor and drop the mutable reference anyways though. Putting it on `Application` makes it hard to access within `Register::push`/`Register::write`. We could put it behind a static (like [this](https://github.com/the-mikedavis/helix/commit/3d44c68d27f2029911187c0119af0c0cdd49decb)) which is a little ugly but eventually we can get rid of almost all references of it by reimplementing all clipboard stuff in terms of registers (like [this](https://github.com/the-mikedavis/helix/commit/59833329e09ba7ce971b498c52e4402670189928)).